File: LeetCode/70.climbing-stairs.cpp
/*
 * @lc app=leetcode id=70 lang=cpp
 *
 * [70] Climbing Stairs
 */

#include<bits/stdc++.h>
using namespace std;
#define ll long long
#define vi vector<int>
#define pii pair<int,int>
#define fr(i,n) for(int i=0;i<n;i++)
#define all(v) v.begin(),v.end()
#define mem(a,b) memset(a,b,sizeof(a))
#define MOD 1000000007
#define pb push_back
#define mp make_pair
#define pii pair<int,int>
#define vll vector<ll>
#define vc vector<char>
#define vs vector<string>
    




// @lc code=start
class Solution {
public:
    int climbStairs(int n) {

        vector<int> dp(n+2,0);

        dp[1]=1;
        dp[2]=2;

        for(int i=3;i<=n;i++){

            dp[i]=dp[i-1]+dp[i-2];

        }

        return dp[n];
        
    }
};
// @lc code=end

File: LeetCode/70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
class Solution:



  


    def climbStairs(self, n: int) -> int:



        1. Recursion
        if n == 1:
            return 1
        if n == 2:
            return 2
        return self.climbStairs(n-1) + self.climbStairs(n-2)

        # 2. Recursion with memoization
        memo = {}
        def helper(n):
            if n in memo:
                return memo[n]
            if n == 1:
                return 1
            if n == 2:
                return 2
            memo[n] = helper(n-1) + helper(n-2)
            return memo[n]
        return helper(n)

        # 3. Iteration
        if n == 1:
            return 1
        if n == 2:
            return 2
        a, b = 1, 2
        for i in range(3, n+1):
            a, b = b, a+b
        return b



        
# @lc code=end

